geochem_inverse_optimize.py

```diff
@@ -267,20 +319,145 @@ def get_downstream_prediction_dictionary(self) -> ElementData:
         return predictions
 
     def get_upstream_prediction_dictionary(self) -> ElementData:
+        if self.continuous:
+            raise Exception(
+                "Warning: `get_upstream_prediction_dictionary` only valid for discrete networks"
+            )
         # Get the predicted upstream concentration we found
         predictions: ElementData = {}
         for sample_name, data in self.sample_network.nodes(data=True):
             data = data["data"]
             predictions[sample_name] = data.my_value.value
         return predictions
 
+    def get_upstream_prediction_map(self) -> np.array:
+        if not self.continuous:
+            raise Exception(
+                "Warning: `get_upstream_prediction_map` only valid for continuous networks"
+            )
+        out = np.zeros(self.grid.area_labels.shape)
+        xstep = out.shape[1] / self.grid.nx
+        ystep = out.shape[0] / self.grid.ny
+        # Loop through inversion grid nodes
+        for i in range(self.grid.nx):
+            for j in range(self.grid.ny):
+                # indices which subdivide the areas on the base array for each inversion grid.
+                x_start = int(i * xstep)
+                x_end = int((i + 1) * xstep)
+                y_start = int((j * ystep))
+                y_end = int((j + 1) * ystep)
+                node = self.grid.node_arr[j, i]
+                # Catch exception for nodes outside of area
+                if node.concentration:
+                    val = node.concentration.value
+                else:
+                    val = np.nan
+                out[y_start:y_end, x_start:x_end] = val
+        return out
+
     def get_misfit(self) -> float:
         return cp.norm(cp.vstack(self._primary_terms)).value
 
     def get_roughness(self) -> float:
         return cp.norm(cp.vstack(self._regularizer_terms)).value
 
 
+@dataclass
+class InverseNode:
+    """A single node on an inversion grid.
+
+    Each `InverseNode` corresponds to a rectangle of pixels on the base raster
+    and is associated with a single sample site. The sample site which each
+    `InverseNode` is associated to is based on sub-catchment which the *centre*
+    of the rectangle lies in. For rectangles which overlap two catchments, there
+    may be some inaccuracies as each `InverseNode` can only be associated with
+    one sample site. As the resolution increases this inaccuracy decreases.
+
+    Attributes:
+        left_neighbour (InverseNode): Pointer to left-neighbouring InverseNode
+        top_neighbour (InverseNode): Pointer to vertically-above InverseNode
+        sample_num (string): Samplename associated with this node
+    """
+
+    concentration: cp.Variable = None
```

Comment:
If the default value is `None`, these all need to be qualified as `Optional`. Avoid default values here if you can because robust code will need to sanity check them before use and it's nice to avoid that complexity.

Reply:
Removed default values